diffrax/_adjoint.py

```diff
@@ -852,3 +865,427 @@ def loop(
+    method.
+
+    This method automatically wraps the passed solver to create an algebraically
+    reversible version of that solver. In doing so, gradient calculation is exact
+    (up to floating point errors) and backpropagation becomes a linear in time $O(n)$
+    and constant in memory $O(1)$ algorithm in the number of steps $n$.
+
+    The reversible adjoint can be used when solving ODEs/CDEs/SDEs and is
+    compatible with any [`diffrax.AbstractSolver`][]. Adaptive step sizes are also
+    supported.
+
+    !!! note
+
+        This adjoint can be less numerically stable than
+        [`diffrax.RecursiveCheckpointAdjoint`][] and [`diffrax.DirectAdjoint`][].
+        Stability can be largely improved by using [double (64bit) precision](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#double-64bit-precision)
+        and [smaller/adaptive step sizes](https://docs.kidger.site/diffrax/api/stepsize_controller/).
+
+    ??? cite "References"
+
+        This algorithm was developed in:
+
+        ```bibtex
+        @article{mccallum2024efficient,
+            title={Efficient, Accurate and Stable Gradients for Neural ODEs},
+            author={McCallum, Sam and Foster, James},
+            journal={arXiv preprint arXiv:2410.11648},
+            year={2024}
+        }
+        ```
+
+        And built on previous work by:
+
+        ```bibtex
+        @article{kidger2021efficient,
+            title={Efficient and accurate gradients for neural sdes},
+            author={Kidger, Patrick and Foster, James and Li, Xuechen Chen and Lyons,
+                    Terry},
+            journal={Advances in Neural Information Processing Systems},
+            volume={34},
+            pages={18747--18761},
+            year={2021}
+        }
+        ```
+
+        ```bibtex
+        @article{zhuang2021mali,
+            title={Mali: A memory efficient and reverse accurate integrator for neural
+                    odes},
+            author={Zhuang, Juntang and Dvornek, Nicha C and Tatikonda, Sekhar and
+            Duncan, James S},
+            journal={arXiv preprint arXiv:2102.04668},
+            year={2021}
+        }
+        ```
+    """
+
+    l: float = 0.999
+
+    def loop(
+        self,
+        *,
+        args,
+        terms,
+        solver,
+        saveat,
+        init_state,
+        passed_solver_state,
+        passed_controller_state,
+        event,
+        **kwargs,
+    ):
+        # `is` check because this may return a Tracer from SaveAt(ts=<array>)
+        if (
+            eqx.tree_equal(saveat, SaveAt(t1=True)) is not True
+            and eqx.tree_equal(saveat, SaveAt(steps=True)) is not True
+            and eqx.tree_equal(saveat, SaveAt(t0=True, steps=True)) is not True
+        ):
+            raise ValueError(
+                "Can only use `diffrax.ReversibleAdjoint` with "
+                "`saveat=SaveAt(t1=True)` or `saveat=SaveAt(steps=True)`."
+            )
```

Comment:
I'm guessing that `SaveAt(steps=True, t1=True)` or `SaveAt(t0=True, steps=True, t1=True)` should also be allowed?